Redesign of events page

hedyhli opened the pull request:
Closes #494 

This PR attempts address the proposal from the linked issue by redesigning our events page to draw more engagement. This includes making the content more easily accessible and organized.

It is also an opportunity to update the page to be more up-to-date on the latest events for this new year.

Pages on preview:
https://deploy-preview-1210--pydis-static.netlify.app/events/

---

Behold, responsive masonry in CSS only

https://github.com/python-discord/site/assets/50042066/583bc66d-5741-43a6-ad1b-e8dcda1d9f32

---

## Implications

The events page would need to be updated more frequently due to this redesign. This means every time an event starts or finishes, we should update the page.

## Limitations

### Masonry grids

To have a good-looking grid with items of variable heights for our regular and one-off events, masonry was required.

<details><summary>Why do we want Masonry??</summary>
<blockquote>

It isn't easy to have all our event descriptions to have the same visual height.

<img width="652" alt="image" src="https://github.com/python-discord/site/assets/50042066/3c34bcd6-4dcd-4e21-b2a8-793a6d144e49">

It might be possible to have flexible thumbnail heights though, to compensate the offsets created from the variable description heights. However, to have the Bulma card component sections work with flexbox is not very easy and I found Masonry to be more flexible.

</blockquote>
</details>

For maximum end-user-experience and developer experience, [Masonry itself](https://masonry.desandro.com/) would be probably be the best bet. However I was against adding a whole javascript dependency just for these two sections so I made use of the [newly discovered CSS-only flexbox solution](https://tobiasahlin.com/blog/masonry-with-css/) instead.

We can't simply use Bulma columns because we need to preserve item ordering on mobile, where no columns are used.

<details><summary>What do you mean??</summary>
<blockquote>

If we use bulma columns, we will define a div with class "column" for each visual column. On wide viewports it will show like this:

<img width="320" alt="image" src="https://github.com/python-discord/site/assets/50042066/d4fdfe20-e4ab-4dd1-84ce-3a70ea8b741e">

But on mobile, where we disable side-by-side columns it will show like this:

<img width="111" alt="image" src="https://github.com/python-discord/site/assets/50042066/913ccc5a-73eb-44bd-b26b-4f0cb263d88d">

<small><a href="https://excalidraw.com/#json=13yf2u-vTLnqdLU9blJpy,b77kg5brRAyKPVbeftMyAA">🔗 Excalidraw</a></small> 

This is not desirable for the regular events grid because we might want to have a specific ordering for the showcase, nor the one-off events because we might want it to be in reverse-chronological order.

</blockquote>
</details>

This came with the limitation that whenever items in the grids are updated we have to modify the `height` in CSS manually, ensuring it has enough room for all screen widths from around 600px to 1000px. This was a pain during development, and for the future when we are adding/removing items from the grid the person who updates it should not be overly involved in the CSS.

The implementation is in the `events/base.css` CSS file with comments.

## Ideas for the future

- #1269 

## Assets

Should use the same icons for the same event, or should we use the annual branding? For instance, we have different branding for each Pyweek and Code Jam, should the icon used in the "scheduled"/"previous" events use the specific branding for that year, this makes multiple entries of the same event have different icons.

Currently, I use the same Pyweek icons, but for Code Jam 2023, I used the 2023 icon, and for the rest, used 2022 icon.

## Todo

Design
- [x] Implement masonry layout, hopefully without JS dependency
- [x] Consider using more fancy colors than white
- [x] Fix up spacing
- [x] Fix layout on tablets
- [x] Ensure consistency on mobile
- [x] Reconsider better labels and title text

Content
- [x] Add one-off events list
- [x] Apply changes suggested by Kat (on owners qna)
- ~~Finalize the assets to use~~
- [x] Finalize list and order of events
- [x] Ensure event schedules are correct
- [x] Fill up all the event pages

Etc.
- [x] Do dark theme stuff

--- Optimistic deadline: February ---
--- WIP ➡️ ready for reviews ---

Maybe
- [x] Minify and/or lazy-load all the images
- [x] Document in a readme how to update info in the future for new events
- Should the events page get a special dazzling hero section similar to the home page with a confetti icon?

Commit:
Events: Initial redesign

- Added standalone pages for each event
- Added initial design of new layout
- Refactored some sidebars and templates to update current/upcoming
  event
- Put two CJ10 assets into their own dedicated folder

Includes dummy text for event descriptions and content that is not
up-to-date: Current and Upcoming event components.

Commit:
Events: Update content and remove fillers

Commit:
Events: Consistent Py*W*eek capitalization and better sidebar

Commit:
Events: Fix title sizing and hidden-on-mobile group

Commit:
Events: Use Masonry layout for gallery of events

Commit:
Events: Use nice cover images with backgrounds

Commit:
Events: Revert height change

Commit:
Events: Make good use of bulma components

Commit:
Merge branch 'main' into feat/events-redesign

Commit:
Events: Adjust spacing

Commit:
Merge branch 'feat/events-redesign' of github.com:python-discord/site into feat/events-redesign

Commit:
Events: Add one-time events, fix elements

Commit:
Events: Adjust date style & add more events

Commit:
Events: Maybe fix masonry grid on tablets

Commit:
Events: Re-organize templates and significantly improve layouts

Aka, many minor changes I shouldn't've grouped together in a single
commit.

- Add main page events description
- Use date icons every where
- Fill up one-off events list
- Use a media-like layout for one-off events instead
- Use 2-column masonry for one-off events list
- Update schedules for this year
- Add a colored top border line to cards in the main gallery that do not
  have thumbnail banners
- Add april fools to fill the gap in the second row so the grid looks
  better
- Re-purpose 'all events' sidebar as 'ToC' with appropriate anchor links
- Use circle-info icon everywhere for learn more buttons
- Added python release streams to the one-off events

That light blue cloud icon for devops hour looks dope IMO :P

- And of course, thanks to the layout change for one-off events list I
  do not have to hurriedly create ad-hoc thumbnails for them 😔. phew
- Which means our fancy hats workshop can use the beautiful magician hat
  emoji as the icon!

- Renamed 'scheduled' and 'recent' events to be year-independent. This
  makes it consistent because we might end up updating the list
  mid-year. This means all dates will have the year specified.

The pain of updating the gallery sections' height for both main and
one-off events section continues to be a limitation.

Commit:
Events: Adjust phrasing in description

Commit:
Fix conflicts

Commit:
Events: Fix grammars and add 2023 code jam logo

Commit:
Events: Update the events to include in galleries

Commit:
Merge branch 'main' into feat/events-redesign

Commit:
Fix events page for dark mode

Commit:
Events: Lazy load larger images

Commit:
Dark: Give more contrast to grey text

Commit:
Events: Remove unnecessary explicit icon color

Commit:
Events: Add draft april fools description

Commit:
Events: Remove assets that are no longer needed

Commit:
Add READMEs to apps/events and templates/events

Commit:
Merge branch 'main' into feat/events-redesign

janine9vn reviewed (approved):
I didn't take an in-depth look at the code because I more-so care how it presents (and html is only so interesting). Overall, I love it! It presents what we do really well and is really and easily digestible.

Thank you so much for your work on this.

Commit:
Merge branch 'main' into feat/events-redesign

Commit:
Events: Update PyWeek 37 to show as ongoing event

Commit:
Merge branch 'main' into feat/events-redesign

Commit:
Events: Update PyWeek 37 from Ongoing to Recent event

It's not completely over yet as the winners are yet to be announced, but
the coding phase, as shown in the date "March 17-24th" has passed so it
no longer makes sense to list it as ongoing.

jchristgit commented on pydis_site/templates/events/other_events.html:
```diff
@@ -0,0 +1,159 @@
+                            <span>Oct. 25, 2022</span>
+                        </span>
+                    </p>
+                    <p>Pablo Galindo, the 3.11 Python Release Manager and CPython Core Developer, walked through the release process and g number of guests joined to talk about the exciting new features to be found in 3.11. At the end of the release we'll also hosted a Q&A session.</p>
```
```suggestion
                    <p>Pablo Galindo, the 3.11 Python Release Manager and CPython Core Developer, walked through the release process and a number of guests joined to talk about the exciting new features to be found in 3.11. At the end of the release we also hosted a Q&A session.</p>
```

jchristgit commented on pydis_site/templates/events/pages/game-jams/_index.html:
```diff
@@ -8,5 +8,25 @@
+        <p>
+            The participants will have ten days to create a game using the technology we've selected, and drawing
+            inspiration from a provided theme. After the event, a panel of judges will play all the games and select a
+            winner. The top 5 will featured in a special video on our <a href="https://www.youtube.com/channel/UCQsrA4xo6jvdgsJZhKaBL6w" target="_blank" rel="noopener">YouTube channel</a>.
```
```suggestion
            winner. The top 5 will be featured in a special video on our <a href="https://www.youtube.com/channel/UCQsrA4xo6jvdgsJZhKaBL6w" target="_blank" rel="noopener">YouTube channel</a>.
```

hedyhli replied:
This too, our current page sure has some grammatical errors. Thanks!

jchristgit commented on pydis_site/templates/events/pages/pyweek/_index.html:
```diff
@@ -0,0 +1,30 @@
+{% block event_content %}
+<p>
+    For the past 15 years, <a href="https://pyweek.org" target="_blank" rel="noopener">PyWeek</a> has been running a bi-annual game jam for the
+    Python language. As of 2020, we are excited to say we are officially partnered with PyWeek to co-run these
```
Hmm, I'm not sure if this is necessary from a grammatical perspective, but I feel it reads more smoothly this way:
```suggestion
    Python language. As of 2020, we are excited to say that we are officially partnered with PyWeek to co-run these
```
What do you think?

hedyhli replied:
This was copied from the current description on our events page, but I agree! It's a good suggestion -- it sounds better.

hedyhli replied:
ece9c315

Also fixed another instance of this content on the annual events gallery.

jchristgit reviewed (changes requested):
Looks very good to me, thank you for the hard work!

I have a few tiny nitpicks, after that it looks good to merge to me. Thank you again!

Commit:
Events: Fix grammar in content

Co-authored-by: jchristgit <jc@jchri.st>
